Promotes Source IP preservation for Virtual IPs from Beta to GA

MrHohn opened the pull request:
Fixes #33625. Feature issue: kubernetes/features#27.

Bullet points:
- Declare 2 fields (ExternalTraffic and HealthCheckNodePort) that mirror the ESIPP annotations.
- ESIPP alpha annotations will be ignored.
- Existing ESIPP beta annotations will still be fully supported.
- Allow promoting beta annotations to first class fields or reversely.
- Disallow setting invalid ExternalTraffic and HealthCheckNodePort on services. Default ExternalTraffic field for nodePort or loadBalancer type service to "Global" if not set.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Promotes Source IP preservation for Virtual IPs to GA.

Two api fields are defined correspondingly:
- Service.Spec.ExternalTrafficPolicy <- 'service.beta.kubernetes.io/external-traffic' annotation.
- Service.Spec.HealthCheckNodePort <- 'service.beta.kubernetes.io/healthcheck-nodeport' annotation.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41162)
<!-- Reviewable:end -->

MrHohn commented:
@bprashanth @thockin

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2518,59 +2531,90 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
 
-	_, healthCheckBetaOk := service.Annotations[apiservice.BetaAnnotationHealthCheckNodePort]
-	_, onlyLocalBetaOk := service.Annotations[apiservice.BetaAnnotationExternalTraffic]
+type serviceExternalTrafficAnnotationsUpdateStatus struct {
```
comments?

MrHohn replied:
Done.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -2518,59 +2531,90 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
+	// to first class fields before making any modifications, even though the
+	// system continues to respect the beta version.
+
+	// On upgrading to a 1.6 cluster, the user is locked in at the current
```
I'm not sure I get this.  Why do we need to stop users from creating new instances with the beta annotation - this is a breaking change, and I don't think we can allow it.  Beta has to keep working for at least 1 release, and that means fully working.

MrHohn replied:
I wanted to say up-conversion, but saw [your old comment about it](https://github.com/kubernetes/kubernetes/pull/33957#discussion_r82726939). Supporting beta annotations for one more release seems to be the right thing to do. I will re-think this part.

thockin replied:
Yeah, we should not up-convert

MrHohn commented on pkg/api/types.go:
```diff
@@ -2353,6 +2363,20 @@ type ServiceSpec struct {
+	// local endpoints only. This preserves Source IP and avoids a second hop for
+	// LoadBalancer and Nodeport type services.
+	// +optional
+	ExternalTraffic ServiceExternalTrafficType
```
Sounds good, so the candidates would be:
```
externalTrafficPolicy: Local
externalTrafficPolicy: Global
```

And do we want to use `externalTrafficPolicy: None` for the other service types (not NodePort or LoadBalancer)? This field is now being defaulted to empty string ("").

thockin replied:
I guess "" is appropriate for those cases.  Defaulting (and update) should detect LoadBalancer and NodePort and choose "Global", I think.  Thoughts?  If we do that then we really should change it back to "" if the type is updated to anything else.

MrHohn replied:
Yeah I have the same thought.
[function defaultExternalTrafficIfNeeded()](https://github.com/kubernetes/kubernetes/pull/41162/commits/744766dd401f8f1fcbafd57142a91938ee5bcfef#diff-8b1ee383fe1e9f26f76c775a274ec900R553) exists for defaulting this field to Global for LoadBalancer and NodePort service if this field is not set.
And [function externalTrafficUpdate()](https://github.com/kubernetes/kubernetes/pull/41162/commits/744766dd401f8f1fcbafd57142a91938ee5bcfef#diff-8b1ee383fe1e9f26f76c775a274ec900R271) exists for changing this field back to "" when the type is updated to others (not LoadBalancer nor NodePort).

MrHohn replied:
Changed to use `externalTrafficPolicy: Local or Global`.

MrHohn commented on pkg/api/validation/validation.go:
```diff
@@ -2477,6 +2477,19 @@ func validateServiceFields(service *api.Service) field.ErrorList {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("healthCheckNodePort"), service.Spec.HealthCheckNodePort,
+			"can not set HealthCheckNodePort to negative value"))
+	}
+	if service.Spec.ExternalTraffic != "" &&
```
Yeah I also think this is inappropriate. I'm changing the default value to None.

MrHohn replied:
Updates...[Per discussion above](https://github.com/kubernetes/kubernetes/pull/41162#discussion_r102340291), "" might be allowed for service that does not have LoadBalancer or NodePort type.

thockin commented:
@bprashanth  - when this went Beta I raised an issue wrt healthchecks and NodePort.  I can't find the comment - do you recall what the issue I raised was?

MrHohn commented on pkg/api/service/util.go:
```diff
@@ -66,3 +68,49 @@ func GetLoadBalancerSourceRanges(service *api.Service) (netsets.IPNet, error) {
+	// If they transition to the first class field, there's no way to go back to beta without
+	// rolling back the cluster.
+	if l, ok := service.Annotations[BetaAnnotationExternalTraffic]; ok {
+		if l == string(api.ServiceExternalTrafficTypeOnlyLocal) {
```
Done.

MrHohn commented on pkg/api/validation/validation.go:
```diff
@@ -2477,6 +2477,19 @@ func validateServiceFields(service *api.Service) field.ErrorList {
+	// Validate ExternalTraffic and HealthCheckNodePort have legal value.
+	if service.Spec.HealthCheckNodePort < 0 {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("healthCheckNodePort"), service.Spec.HealthCheckNodePort,
+			"can not set HealthCheckNodePort to negative value"))
```
Done.

MrHohn commented on pkg/api/validation/validation.go:
```diff
@@ -2518,59 +2531,90 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
+		(service.Spec.Type != api.ServiceTypeLoadBalancer ||
+			service.Spec.ExternalTraffic != api.ServiceExternalTrafficTypeOnlyLocal) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "healthCheckNodePort"), service.Spec.ExternalTraffic,
+			"can only be set on loadBalancer service with ExternalTraffic=OnlyLocal"))
```
Done.

MrHohn commented:
Per [discussion above](https://github.com/kubernetes/kubernetes/pull/41162#discussion_r102363543), updated to fully support both ESIPP Beta annotations and first class fields:
- Allow legally create Beta annotations or first class fields.
- Allow legally modify Beta annotations or first class fields.
- Allow legally delete Beta annotations or reset first class fields.
- Allow transition from Beta annotations to first class fields or reverse.
- Disallow mixing Beta annotations with first class fields.

For the ease of review, quick summarize service creation/update path:
- Creation:
  1. `validateServiceExternalTrafficFields()` -- Validate values are valid and not mixing Beta with GA.
  2.  `DefaultExternalTrafficPolicyIfNeeded()` -- Default ExternalTrafficPolicy based on Service Type.
  3.  `ValidateServiceExternalTrafficFieldsBeforeAction()` -- Validate the combination of ExternalTrafficPolicy, HealthCheckNodePort and Service Type.
- Update:
  1. `validateServiceExternalTrafficFields()` -- Validate values are valid and not mixing Beta with GA.
  2. `healthCheckNodePortUpdate()` -- Handle HealthCheckNodePort allocation/release and adjust HealthCheckNodePort.
  3. `externalTrafficPolicyUpdate()` -- Adjust ExternalTrafficPolicy for the ease of Service Types transition.
  4.  `ValidateServiceExternalTrafficFieldsBeforeAction()` -- Validate the combination of ExternalTrafficPolicy, HealthCheckNodePort and Service Type.

Also split the validation+registry unit tests commit out and hopefully make it easier to review.

MrHohn commented:
cc @freehan who might also be familiar with ESIPP validation :)

MrHohn commented:
@k8s-bot gce etcd3 e2e test this

MrHohn commented:
@k8s-bot non-cri e2e test this

@k8s-bot non-cri node e2e test this

MrHohn commented:
Quick update:
This PR is standing by until we fully understand what impact kubernetes/ingress#19 would have on onlylocal services.

ethernetdan commented:
Removing issue from 1.6 milestone since code freeze has begun:
* If this is a bug fix, please label as `kind/bug` and reapply the milestone.
* If this is a feature for 1.6, an [exception](https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md) should be filed for it before 11:00PM PST on 3/2. Note that most exceptions to code freeze are not accepted.
* If this is a feature for 1.7, feel free to add the 1.7 milestone to the PR.

/cc @kubernetes/release-team @kubernetes/kubernetes-release-managers

calebamiles commented:
Moving this PR to the 1.7 milestone @MrHohn.

**cc: @ethernetdan, @thockin, @kubernetes/kubernetes-release-managers**

thockin commented:
rebase and ping?

MrHohn committed:
Define ESIPP GA fields and deprecate alpha annotations

MrHohn committed:
Update ESIPP validation and service registry unit tests

MrHohn commented:
@thockin Sorry for the delay, PR is rebased and ready.

MrHohn committed:
Change to use ESIPP first class fields and update comments

MrHohn committed:
Autogenerated files

MrHohn commented:
@k8s-bot gce etcd3 e2e test this

MrHohn commented:
Rebased.

Note that most of the worth-looking-at changes are in the first two commits :)

MrHohn commented:
/assign @freehan 

Many thanks for helping out!

MrHohn commented:
Rebased :)

@freehan Any changes to take a look?

MrHohn commented:
@k8s-bot kops aws e2e test this, issue: #43236

freehan commented on pkg/api/service/util.go:
```diff
@@ -103,21 +102,59 @@ func NeedsHealthCheck(service *api.Service) bool {
-				continue
-			}
-			return int32(p)
+	// First check the beta annotation and then the first class field. This is so
```
nit: s/This is so/so that

MrHohn replied:
Done.

freehan commented on pkg/api/service/util.go:
```diff
@@ -103,21 +102,59 @@ func NeedsHealthCheck(service *api.Service) bool {
+
+// ClearExternalTrafficPolicy resets the ExternalTrafficPolicy field.
+func ClearExternalTrafficPolicy(service *api.Service) {
+	// First check the beta annotation and then the first class field. This is so existing
```
same here

MrHohn replied:
Done.

freehan commented on pkg/api/service/util.go:
```diff
@@ -103,21 +102,59 @@ func NeedsHealthCheck(service *api.Service) bool {
+// SetServiceHealthCheckNodePort sets the given health check node port on service.
+// It does not check whether this service needs healthCheckNodePort.
+func SetServiceHealthCheckNodePort(service *api.Service, hcNodePort int32) {
+	// First check the beta annotation and then the first class field. This is so existing
```
same here

MrHohn replied:
Done.

freehan commented on pkg/api/service/util.go:
```diff
@@ -103,21 +102,59 @@ func NeedsHealthCheck(service *api.Service) bool {
+// DefaultExternalTrafficPolicyIfNeeded defaults the ExternalTrafficPolicy field
+// for NodePort / LoadBalancer service to Global for consistency.
+// TODO: Move this default logic to default.go once beta annotation is deprecated.
+func DefaultExternalTrafficPolicyIfNeeded(service *api.Service) {
```
nit: s/DefaultExternalTrafficPolicyIfNeeded/SetDefaultExternalTrafficPolicyIfNeeded

MrHohn replied:
Done.

freehan commented on pkg/api/v1/service/util.go:
```diff
@@ -103,23 +102,61 @@ func NeedsHealthCheck(service *v1.Service) bool {
+// DefaultExternalTrafficPolicyIfNeeded defaults the ExternalTrafficPolicy field
+// for NodePort / LoadBalancer service to Global for consistency.
+// TODO: Move this default logic to default.go once beta annotation is deprecated.
+func DefaultExternalTrafficPolicyIfNeeded(service *v1.Service) {
```
nit: s/DefaultExternalTrafficPolicyIfNeeded/SetDefaultExternalTrafficPolicyIfNeeded

MrHohn replied:
Done.

freehan commented on pkg/api/validation/validation.go:
```diff
@@ -2785,61 +2786,68 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
+// validateServiceExternalTrafficFieldsValue validates ExternalTraffic related annotations
 // have legal value.
-func validateServiceExternalTrafficFields(service *api.Service) field.ErrorList {
+func validateServiceExternalTrafficFieldsValue(service *api.Service) field.ErrorList {
```
Not sure if it is relevant. 
Is it allowed to use healthcheck node port with Global?
Is it allowed to mixa usage of beta annotation and fields? (I think annotaion is preempting, so we do not care? )

MrHohn replied:
> Is it allowed to use healthcheck node port with Global?

Nope, we disallow this case in ValidateServiceExternalTrafficFieldsCombination().

> Is it allowed to mixa usage of beta annotation and fields? (I think annotaion is preempting, so we do not care? )

Nope, we disallow this case in validateServiceExternalTrafficAPIVersion().

Sorry for the mess, but I have to separate the validation into multiple pieces because they are called in different phases --- especially ValidateServiceExternalTrafficFieldsCombination() needs to be called after defaulting and resetting ESIPP fields :(

freehan reviewed (commented):
Looks good overall. Some nits.

MrHohn reviewed (commented):
Thanks! WIll address the comments soon.

freehan commented:
/lgtm

MrHohn commented:
Thanks! Squashing commits.

MrHohn committed:
Adjust ESIPP validation and service registry codes
